sessionctx/variable/noop.go

```diff
@@ -23,6 +23,39 @@ import (
 // but changing them has no effect on behavior.
 
 var noopSysVars = []*SysVar{
+	// It is unsafe to pretend that any variation of "read only" is enabled when the server
+	// does not support it. It is possible that these features will be supported in future,
+	// but until then...
+	{Scope: ScopeGlobal | ScopeSession, Name: TxReadOnly, Value: BoolOff, Type: TypeBool, Validation: func(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
+		if TiDBOptOn(normalizedValue) && !vars.EnableNoopFuncs {
+			return BoolOff, ErrFunctionsNoopImpl.GenWithStackByArgs("READ ONLY")
+		}
+		return normalizedValue, nil
+	}},
+	{Scope: ScopeGlobal | ScopeSession, Name: TransactionReadOnly, Value: BoolOff, Type: TypeBool, Validation: func(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
+		if TiDBOptOn(normalizedValue) && !vars.EnableNoopFuncs {
+			return BoolOff, ErrFunctionsNoopImpl.GenWithStackByArgs("READ ONLY")
+		}
+		return normalizedValue, nil
+	}},
```

Comment:
Is there only 1 or 2 copies of data for these 2 variables?

Reply:
There are two copies kept, but the synonym check in `executor/set.go` ensures both are updated:
```
func (e *SetExecutor) getSynonyms(varName string) []string {
	synonyms, ok := variable.SynonymsSysVariables[varName]
	if ok {
		return synonyms
	}

	synonyms = []string{varName}
	return synonyms
}
```

This check should not technically be in `executor/set.go` though, since it's possible the sysvar could be updated externally. I will fix it in a future issue.

Having both sysVars check for validity is required because if the validation function is specified only for `TransactionReadOnly`, `getSynonyms` would first change the value of `TxReadOnly` and then fail when trying to update `TransactionReadOnly`. i.e. there would be no atomic semantic in updating both. This is because `sessionctx/variable/sysvar.go:addSynonymsSysVariables()` does not reorder the synonyms for consistent ordinality.